File: tests/test_to_srt.py
import os
import re
import unittest

from to_srt import to_srt


class TestToSrt(unittest.TestCase):
    SAMPLES_DIR = os.path.join(os.path.dirname(__file__), "..", "samples")

    def test_sample_files(self):
        input_extensions = {".xml", ".vtt"}
        for input_file in sorted(os.listdir(self.SAMPLES_DIR)):
            base, ext = os.path.splitext(input_file)
            if ext.lower() not in input_extensions:
                continue
            with self.subTest(file=input_file):
                expected_srt = os.path.join(self.SAMPLES_DIR, base + ".srt")
                input_path = os.path.join(self.SAMPLES_DIR, input_file)
                with open(input_path, "r", encoding="utf-8") as f:
                    input_content = f.read()
                with open(expected_srt, "r", encoding="utf-8") as f:
                    expected_content = f.read()
                generated_content = to_srt(input_content, ext, delay_ms=0)
                self.assertEqual(generated_content, expected_content)

    def test_delay_subdirectory(self):
        delay_dir = os.path.join(self.SAMPLES_DIR, "delay")
        delay_pattern = re.compile(r"^(.+?)_(plus|minus)_(\d+)$")

        for f in sorted(os.listdir(delay_dir)):
            base, ext = os.path.splitext(f)
            if ext.lower() != ".srt":
                continue
            match = delay_pattern.match(base)
            if not match:
                continue
            prefix, sign, delay_val = match.groups()
            delay_ms = int(delay_val) if sign == "plus" else -int(delay_val)
            input_ext = None
            for ext in [".vtt", ".xml", ".srt"]:
                if os.path.exists(os.path.join(delay_dir, prefix + ext)):
                    input_file = prefix + ext
                    input_ext = ext
                    break
            if input_ext is None:
                raise FileNotFoundError(
                    f"No input file found for prefix '{prefix}' "
                    f"(looked for .vtt, .xml, .srt)"
                )
            with self.subTest(input_file=input_file, delay=delay_val):
                expected_srt = os.path.join(delay_dir, f)
                input_path = os.path.join(delay_dir, input_file)
                with open(input_path, "r", encoding="utf-8") as f:
                    input_content = f.read()
                with open(expected_srt, "r", encoding="utf-8") as f:
                    expected_content = f.read()
                generated_content = to_srt(input_content, input_ext, delay_ms=delay_ms)
                self.assertEqual(generated_content, expected_content)


if __name__ == "__main__":
    unittest.main()

File: tests/test_to_srt.js
const { readdir, readFile } = require('fs/promises');
const { resolve, join } = require('path');
const { test, describe } = require('node:test');
const assert = require('node:assert');

const { toSrt } = require('../web/script.js');

const SAMPLES_DIR = resolve(__dirname, '..', 'samples');

describe('toSrt conversion tests', () => {
    test('sample files in samples/ directory', async () => {
        const inputExtensions = new Set(['.xml', '.vtt']);
        const files = await readdir(SAMPLES_DIR);
        const sortedFiles = files.sort();

        for (const inputFile of sortedFiles) {
            const base = inputFile.substring(0, inputFile.lastIndexOf('.'));
            const ext = inputFile.substring(inputFile.lastIndexOf('.')).toLowerCase();

            if (!inputExtensions.has(ext)) continue;
            if (inputFile.includes('ep2')) continue;

            const expectedSrt = join(SAMPLES_DIR, base + '.srt');
            const inputPath = join(SAMPLES_DIR, inputFile);

            const inputContent = await readFile(inputPath, 'utf-8');
            const expectedContent = await readFile(expectedSrt, 'utf-8');
            const generatedContent = toSrt(inputContent, ext, 0);

            assert.equal(generatedContent, expectedContent, `Failed for ${inputFile}`);
        }
    });
});

describe('toSrt delay tests', () => {
    test('delay subdirectory', async () => {
        const delayDir = join(SAMPLES_DIR, 'delay');
        const delayPattern = /^(.+?)_(plus|minus)_(\d+)\.srt$/;
        const files = await readdir(delayDir);
        const sortedFiles = files.sort();

        const delayFiles = sortedFiles.filter(f => delayPattern.test(f));
        assert.ok(delayFiles.length > 0, 'No delay test files found');

        for (const f of delayFiles) {
            const match = f.match(delayPattern);
            assert.ok(match, `Filename "${f}" does not match delay pattern`);

            const [, prefix, sign, delayVal] = match;
            const delayMs = sign === 'plus' ? parseInt(delayVal, 10) : -parseInt(delayVal, 10);

            let inputExt = null;
            for (const ext of ['.vtt', '.xml', '.srt']) {
                try {
                    await readFile(join(delayDir, prefix + ext), 'utf-8');
                    inputExt = ext;
                    break;
                } catch {
                    continue;
                }
            }

            assert.ok(inputExt, `No input file found for prefix "${prefix}" (looked for .vtt, .xml, .srt)`);

            const expectedSrt = join(delayDir, f);
            const inputPath = join(delayDir, prefix + inputExt);

            const inputContent = await readFile(inputPath, 'utf-8');
            const expectedContent = await readFile(expectedSrt, 'utf-8');
            const generatedContent = toSrt(inputContent, inputExt, delayMs);

            assert.equal(generatedContent, expectedContent, `Failed for ${f} with delay ${delayMs}`);
        }
    });
});
